### DISCUSSION TITLE
Be able to pass non JSON-serializeable data to my page through `getStaticProps`

### DISCUSSION THREAD

**Discussion by lffg** [vercel/next.js#11807] [Help]
# Feature request

It would be wonderful if we were able to pass non JSON-serializeable data to Next.js pages through `getStaticProps`.

For example, what if I want to pass a JSX component to my page using `getStaticProps`? Something like that:

```ts
import { GetStaticProps } from 'next';
import React from 'react';

export interface Props {
  comp: React.ReactNode;
}

export default function Test(props: Props) {
  return (
    <div>
      <h1>Component:</h1>
      {props.comp}
    </div>
  );
}

export const getStaticProps: GetStaticProps<Props> = async () => {
  return {
    props: {
      comp: <h1>Hello, world!</h1>
    }
  };
};
```

I **cannot**. Next.js yells at me:

> Error serializing `.comp.$$typeof` returned from `getStaticProps` in "/test". Reason: `symbol` cannot be serialized as JSON. Please only return JSON serializable data types.

## Is your feature request related to a problem? Please describe.

I would not say it is a problem, as passing JSON-only serializeable types is intended by Next.js.

The [docs](https://nextjs.org/docs/basic-features/data-fetching#statically-generates-both-html-and-json) say:

> When a page with `getStaticProps` is pre-rendered at build time, in addition to the page HTML file, Next.js generates a **JSON** file holding the result of running `getStaticProps`.

React.js accepts any kind of JavaScript feature and/or data structures. I don't know why to have this kind of limitation, as it prevents more advanced abstractions which uses more complex JavaScript data structures, such as JSX components or even `Symbol`s, `Set`s or `Map`s as they are not JSON seriable.

So, instead of generating a JSON file, Next.js could generate a JavaScript file. Is this doable?

## Describe alternatives you've considered

I don't know how to solve this problem. I think that it is not even possible without the proposed change (Next.js generate JavaScript files instead of JSON ones).

## Additional context

N/A

**Comment by janus-reith** [comment]
@lffg The file, currently JSON, would store the result of something that already executed. In the case of your proposed JS file and a React component, what would that be, and what would be the general shape of data? The file is just expected to export a JS object?
Would it even work to store things like that React component in a file, and just use it elsewhere outside of that previous context?

Not arguing it's a bad idea or would'nt work, just curious.

**Comment by lffg** [comment]
@janus-reith, I don't actually know how Next.js works internally (never explored its source code to know :P)

But I do know that with `getStaticProps`, alongside of the generated static HTML, Next.js generates also a [static] JSON file, contaning somewhat like the returned value of `getStaticProps`, and I think that is why we can't return values that are not JSON-serializeable.

I think that if Next.js generate a plain old JS file — which, yes, would export a normal JavaScript object or array — instead of the current JSON file, the problem would be solved, and we can used all types of JavaScript-supported data structures, such as `Set`s, `Map`s, `Symbol`s and even React Components!

If the Next.js team and the community embrace this idea, I could willingly contribute with a PR. :)

**Comment by janus-reith** [comment]
I think we could leave nextjs out of the equation:
Just with plain nodejs, what would be your approach to savely store any arbitrary object data which possibly can't be JSON-serialized to a file and import that same object elsewhere?

If you're just looking for a way to pass the function that creates a React component, similar to the way you defined it in some file, then you could also do that already by serializing that function as a string as passing it as valid json, and evaluating that later on. But for most usecases I would guess that there might be an easier way to achieve your goal which is less prone to errors.

**Comment by thobhanifreddy** [comment]
One more real time use case here would be, let's say I am using using [Parse Server](https://parseplatform.org/). Here, queries returns parse object instead of json. I won't able to return parse objects using  `getStaticProps`
